demos/eigenvalues_QG_basinmodes/qgbasinmodes.py.rst

```diff
@@ -154,13 +155,23 @@ We define the Test Function :math:`\phi` and the Trial Function
 To build the weak formulation of our equation we need to build two PETSc
 matrices in the form of a generalized eigenvalue problem,
 :math:`A\psi = \lambda M\psi`. We impose the boundary conditions on the
-mass matrix :math:`M`, since that is where we used integration by parts. ::
+mass matrix :math:`M`, since that is where we used integration by parts.
+To avoid spurious eigenvalues associated with the rows of :math:`M`
+that are modified by the Dirichlet boundary conditions, we create
+submatrices that omit these rows and the correspoing columns. ::
 
    a =  beta*phi*psi.dx(0)*dx
    m = -inner(grad(psi), grad(phi))*dx - F*psi*phi*dx
    petsc_a = assemble(a).M.handle
    petsc_m = assemble(m, bcs=bc).M.handle
 
+   bcnodes = Vcg.dof_dset.lgmap.apply(bc.nodes).astype(pyop2.datatypes.IntType)
+   ISbd = PETSc.IS().createGeneral(bcnodes, comm=petsc_a.comm)
+   s, e = petsc_a.getOwnershipRange()
+   ISin = ISbd.complement(s, e)
```

Comment:
In parallel, `ISbd.complement(s, e)` may rise exception:
```
File PETSc/IS.pyx:234, in petsc4py.PETSc.IS.complement()

Error: error code 62
[0] ISComplement() at /home/zzyang/workspace/firedrake-gcc/firedrake-mini-petsc-complex/src/petsc/src/vec/is/is/utils/iscoloring.c:815
[0] Invalid argument
[0] Index set must be sorted
[1:execute]
```
Maybe need add
```
s, e = petsc_a.getOwnershipRange()
bcnodes = bcnodes[np.logical_and(bcnodes < e, bcnodes >= s)]
bcnodes.sort()
```
to delete the ghost nodes and sort the array before creating the `ISbd`.